src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/GrpcRemoteDownloader.java

```diff
@@ -0,0 +1,183 @@
+  public void download(
+      List<URL> urls,
+      Map<URI, Map<String, String>> authHeaders,
+      com.google.common.base.Optional<Checksum> checksum,
+      String canonicalId,
+      Path destination,
+      ExtendedEventHandler eventHandler,
+      Map<String, String> clientEnv)
+      throws IOException, InterruptedException {
+    final FetchBlobRequest request = newFetchBlobRequest(
+        options.remoteInstanceName, urls, authHeaders, checksum, canonicalId);
+    try {
+      FetchBlobResponse response = retrier.execute(
+          () -> requestCtx.call(
+              () -> fetchBlockingStub().fetchBlob(request)));
+      final Digest blobDigest = response.getBlobDigest();
+
+      retrier.execute(() -> requestCtx.call(() -> {
+        try (OutputStream out = newOutputStream(destination, checksum)) {
+          Utils.getFromFuture(cacheClient.downloadBlob(blobDigest, out));
+        }
+        return null;
+      }));
+    } catch (StatusRuntimeException e) {
+      throw new IOException(e);
+    }
+  }
+
+  static FetchBlobRequest newFetchBlobRequest(
+      String instanceName,
+      List<URL> urls,
+      Map<URI, Map<String, String>> authHeaders,
+      com.google.common.base.Optional<Checksum> checksum,
+      String canonicalId) {
+    FetchBlobRequest.Builder requestBuilder =
+        FetchBlobRequest.newBuilder()
+            .setInstanceName(instanceName);
+    for (URL url : urls) {
+      requestBuilder.addUris(url.toString());
+    }
+    if (checksum.isPresent()) {
+      requestBuilder.addQualifiers(
+        Qualifier.newBuilder()
+            .setName("checksum.sri")
+            .setValue(checksum.get().toSubresourceIntegrity()).build());
+    }
+    if (!Strings.isNullOrEmpty(canonicalId)) {
+      requestBuilder.addQualifiers(
+        Qualifier.newBuilder()
+            .setName("bazel.canonical_id")
+            .setValue(canonicalId).build());
+    }
+    if (!authHeaders.isEmpty()) {
+      requestBuilder.addQualifiers(
+        Qualifier.newBuilder()
+            .setName("bazel.auth_headers")
+            .setValue(authHeadersJson(authHeaders)).build());
+    }
+
+    return requestBuilder.build();
+  }
+
+  private FetchBlockingStub fetchBlockingStub() {
+    return FetchGrpc.newBlockingStub(channel)
+        .withInterceptors(TracingMetadataUtils.attachMetadataFromContextInterceptor())
+        .withInterceptors(TracingMetadataUtils.newDownloaderHeadersInterceptor(options))
+        .withCallCredentials(credentials.orElse(null))
+        .withDeadlineAfter(options.remoteTimeout, TimeUnit.SECONDS);
+  }
+
+  private OutputStream newOutputStream(
+      Path destination,
+      com.google.common.base.Optional<Checksum> checksum) throws IOException {
+    OutputStream out = destination.getOutputStream();
+    if (checksum.isPresent()) {
+      out = new HashOutputStream(out, checksum.get());
+    }
+    return out;
+  }
+
+  private static String authHeadersJson(Map<URI, Map<String, String>> authHeaders) {
```

Comment:
How does this interact with the authentication methods from the remote module? Why do we want this?

Reply:
`ctx.download()` allows arbitrary HTTP headers to be passed to the HTTP library for authenticating with the remote endpoint. This is part of the `netrc` support added in ... v2.0? v2.1? Fairly recently.

This code implements that by passing it along to the remote download server, which is free to use (or ignore) that data. The download server would use it when the place being downloaded from (e.g. example.com) needs special headers such as HTTP Basic Auth.